File: packages/conformance/src/producer.ts
import type {
  AdmissionReasonCode,
  AdmissionReport,
  CheckRegistry,
  EvidenceAdmissionResult,
  ProducerRegistry,
  SubjectReference,
} from '@l9/assurance-contracts';
import {
  admitObservations,
  type AdmissionContext,
  type AdmissionLimits,
  validateObservation,
} from '@l9/assurance-evidence';

export interface ConformanceCaseResult {
  readonly index: number;
  readonly status: 'pass' | 'fail';
  readonly reasons: readonly string[];
}

export interface ProducerConformanceReport {
  readonly producerId: string;
  readonly passed: boolean;
  readonly cases: readonly ConformanceCaseResult[];
  readonly admission?: AdmissionReport;
}

export interface ProducerAdmissionConformanceCase {
  readonly id: string;
  readonly value: unknown;
  readonly expectedStatus: EvidenceAdmissionResult['status'];
  readonly expectedReason?: AdmissionReasonCode;
}

export interface ProducerAdmissionConformanceCaseResult {
  readonly id: string;
  readonly passed: boolean;
  readonly actualStatus: EvidenceAdmissionResult['status'];
  readonly reasonCodes: readonly AdmissionReasonCode[];
}

export interface ProducerAdmissionConformanceReport {
  readonly producerId: string;
  readonly passed: boolean;
  readonly cases: readonly ProducerAdmissionConformanceCaseResult[];
  readonly admission: AdmissionReport;
}

interface ProducerConformanceOptions {
  readonly producerId: string;
  readonly subject: SubjectReference;
  readonly producerRegistry: ProducerRegistry;
  readonly checkRegistry: CheckRegistry;
  readonly receivedAt: string;
  readonly maximumAgeSecondsByCheck?: Readonly<Record<string, number>>;
  readonly limits?: Partial<AdmissionLimits>;
}

export async function runProducerConformance(
  values: readonly unknown[],
  options: ProducerConformanceOptions,
): Promise<ProducerConformanceReport> {
  const cases = values.map((value, index) => {
    const validation = validateObservation(value);
    return {
      index,
      status: validation.valid ? 'pass' : 'fail',
      reasons: validation.errors,
    } as const;
  });
  const admission = await admitObservations(values, admissionContext(options));
  const passed =
    values.length > 0 &&
    cases.every((item) => item.status === 'pass') &&
    admission.rejectedCount === 0 &&
    admission.quarantinedCount === 0;
  return { producerId: options.producerId, passed, cases: Object.freeze(cases), admission };
}

export async function runProducerAdmissionConformanceCases(
  cases: readonly ProducerAdmissionConformanceCase[],
  options: ProducerConformanceOptions,
): Promise<ProducerAdmissionConformanceReport> {
  const admission = await admitObservations(
    cases.map((item) => item.value),
    admissionContext(options),
  );
  const results = cases.map((item, index) => {
    const actual = admission.results[index];
    if (!actual) throw new Error(`Missing admission result for conformance case ${item.id}.`);
    const reasonCodes: readonly AdmissionReasonCode[] = Object.freeze(
      actual.reasons.map((reason) => reason.code as AdmissionReasonCode),
    );
    return {
      id: item.id,
      passed:
        actual.status === item.expectedStatus &&
        (item.expectedReason === undefined || reasonCodes.includes(item.expectedReason)),
      actualStatus: actual.status,
      reasonCodes,
    };
  });
  return {
    producerId: options.producerId,
    passed: results.every((item) => item.passed),
    cases: Object.freeze(results),
    admission,
  };
}

function admissionContext(options: ProducerConformanceOptions): AdmissionContext {
  return {
    subject: options.subject,
    producerRegistry: options.producerRegistry,
    checkRegistry: options.checkRegistry,
    receivedAt: options.receivedAt,
    channel: 'local',
    ...(options.maximumAgeSecondsByCheck
      ? { maximumAgeSecondsByCheck: options.maximumAgeSecondsByCheck }
      : {}),
    ...(options.limits ? { limits: options.limits } : {}),
  };
}

File: src/l9_assurance/conformance/producer.py
from __future__ import annotations

from typing import Any, Mapping, Sequence

from l9_assurance.evidence import admit_observations, validate_observation


def run_producer_conformance(
    values: Sequence[Any],
    *,
    producer_id: str,
    subject: Mapping[str, Any],
    producer_registry: Mapping[str, Any],
    check_registry: Mapping[str, Any],
    received_at: str,
    maximum_age_seconds_by_check: Mapping[str, int] | None = None,
    limits: Mapping[str, int] | None = None,
) -> dict[str, Any]:
    cases: list[dict[str, Any]] = []
    for index, value in enumerate(values):
        validation = validate_observation(value)
        producer_matches = bool(
            validation.observation
            and validation.observation["producer"]["id"] == producer_id
        )
        reasons = list(validation.errors)
        if not producer_matches:
            reasons.append(f"Expected producer {producer_id}.")
        cases.append(
            {
                "index": index,
                "status": "pass" if validation.valid and producer_matches else "fail",
                "reasons": reasons,
            }
        )
    context: dict[str, Any] = {
        "subject": subject,
        "producerRegistry": producer_registry,
        "checkRegistry": check_registry,
        "receivedAt": received_at,
        "channel": "local",
    }
    if maximum_age_seconds_by_check is not None:
        context["maximumAgeSecondsByCheck"] = maximum_age_seconds_by_check
    if limits is not None:
        context["limits"] = limits
    admission = admit_observations(values, context)
    passed = bool(values) and all(item["status"] == "pass" for item in cases) and admission["rejectedCount"] == 0 and admission["quarantinedCount"] == 0
    return {"producerId": producer_id, "passed": passed, "cases": cases, "admission": admission}


def run_producer_admission_cases(
    cases: Sequence[Mapping[str, Any]],
    **options: Any,
) -> dict[str, Any]:
    admission = admit_observations(
        [item["value"] for item in cases],
        {
            "subject": options["subject"],
            "producerRegistry": options["producer_registry"],
            "checkRegistry": options["check_registry"],
            "receivedAt": options["received_at"],
            "channel": "local",
            **({"maximumAgeSecondsByCheck": options["maximum_age_seconds_by_check"]} if options.get("maximum_age_seconds_by_check") else {}),
            **({"limits": options["limits"]} if options.get("limits") else {}),
        },
    )
    results = []
    for index, item in enumerate(cases):
        actual = admission["results"][index]
        codes = [reason["code"] for reason in actual["reasons"]]
        expected_reason = item.get("expectedReason")
        results.append(
            {
                "id": item["id"],
                "passed": actual["status"] == item["expectedStatus"] and (expected_reason is None or expected_reason in codes),
                "actualStatus": actual["status"],
                "reasonCodes": codes,
            }
        )
    return {"producerId": options["producer_id"], "passed": all(item["passed"] for item in results), "cases": results, "admission": admission}
